cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramRuntimeProviderLoader.java

```diff
@@ -87,7 +66,7 @@ public ProgramRuntimeProviderLoader(CConfiguration cConf) {
   @Nullable
   public ProgramRuntimeProvider get(ProgramType programType) {
```

Comment:
You can move this method to the super class. You don't need the `defaultProvider` if the `LoadingCache` have type `LoadingCache<EXTENSION_KEY, AtomicReference<EXTENSION>>` and use `AtomicReference(null)` as a place hold for the key that has no extension supports.